packages/sdk/src/client/client.ts

```diff
@@ -838,95 +856,186 @@ export class Client {
     }
 
     this.conditions[ClientCondition.WatchLoop] = true;
-    return attachment.runWatchLoop(
-      (onDisconnect: () => void): Promise<[WatchStream, AbortController]> => {
-        if (!this.isActive()) {
-          this.conditions[ClientCondition.WatchLoop] = false;
-          return Promise.reject(
-            new YorkieError(
-              Code.ErrClientNotActivated,
-              `${this.key} is not active`,
-            ),
-          );
-        }
+    let lastEventTime: number = Date.now();
+    let currentTimeout: number = reconnectionStreamTimeout;
+    let consecutiveTimeouts: number = 0;
 
-        const ac = new AbortController();
-        const stream = this.rpcClient.watchDocument(
-          {
-            clientId: this.id!,
-            documentId: attachment.docID,
-          },
-          {
-            headers: { 'x-shard-key': `${this.apiKey}/${docKey}` },
-            signal: ac.signal,
-          },
+    logger.info(
+      `[WD] c:"${this.getKey()}" initializing watch loop with base timeout: ${reconnectionStreamTimeout}ms`,
+    );
+
+    const resetEventTimer = () => {
+      const previousEventTime = lastEventTime;
+      lastEventTime = Date.now();
+      logger.debug(
+        `[WD] c:"${this.getKey()}" event received after ${Date.now() - previousEventTime}ms`,
+      );
+
+      // Reset timeout if we're getting regular events
+      if (consecutiveTimeouts > 0) {
+        logger.info(
+          `[WD] c:"${this.getKey()}" resetting timeout from ${currentTimeout}ms to ${reconnectionStreamTimeout}ms after ${consecutiveTimeouts} timeouts`,
         );
+        consecutiveTimeouts = 0;
+        currentTimeout = reconnectionStreamTimeout;
+      }
+    };
 
-        attachment.doc.publish([
-          {
-            type: DocEventType.ConnectionChanged,
-            value: StreamConnectionStatus.Connected,
-          },
-        ]);
-        logger.info(`[WD] c:"${this.getKey()}" watches d:"${docKey}"`);
-
-        return new Promise((resolve, reject) => {
-          const handleStream = async () => {
-            try {
-              for await (const resp of stream) {
-                this.handleWatchDocumentsResponse(attachment, resp);
-
-                // NOTE(hackerwins): When the first response is received, we need to
-                // resolve the promise to notify that the watch stream is ready.
-                if (resp.body.case === 'initialization') {
-                  resolve([stream, ac]);
-                }
+    const calculateNextTimeout = () => {
+      consecutiveTimeouts++;
+      const previousTimeout = currentTimeout;
+      // Apply exponential backoff with a maximum of 5 minutes (300000ms)
+      currentTimeout = Math.min(
+        reconnectionStreamTimeout * Math.pow(2, consecutiveTimeouts - 1),
+        300000,
```

Comment:
Is there a specific reason you set it to 300000 ms? (just curious)

Reply:
> I'm currently testing to determine the optimal timeout value for this mechanism. The values will be finalized before this PR is opened for review.

Same as above.